File: frontend/packages/utils/src/tools/time/convert.ts
// Copyright 2025 Beijing Volcano Engine Technology Co., Ltd. and/or its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import dayjs from 'dayjs';
import { isNumber, pick } from 'lodash-es';
import {
  type ConvertToUnixTimestampProps,
  DEFAULT_TIME_FORMAT_TEMPLATE,
} from './constants';
import {
  canConvertToNumber as canConvertToNumberFromCommon,
  isMillisecondTimestamp,
  isSecondTimestamp,
} from './utils';

/**
 * 将字符串类型的时间转换为Unix时间戳
 * @param time 时间字符串
 * @param format 格式化模板（可选）
 * @returns 对应的Unix时间戳
 */
export const convertToUnixTimestamp = ({
  time,
  format = DEFAULT_TIME_FORMAT_TEMPLATE,
  isToSecondTimestamp = false,
  isToMillSecondTimestamp = false,
}: ConvertToUnixTimestampProps): number => {
  let unixTimestamp = time;
  if (typeof time === 'string') {
    if (format) {
      unixTimestamp = dayjs(time, format).valueOf();
    } else {
      unixTimestamp = dayjs(time).valueOf();
    }
  }
  if (
    isToSecondTimestamp &&
    typeof unixTimestamp === 'number' &&
    isMillisecondTimestamp(unixTimestamp)
  ) {
    unixTimestamp = Math.floor(unixTimestamp / 1000);
  }

  if (
    isToMillSecondTimestamp &&
    typeof unixTimestamp === 'number' &&
    isSecondTimestamp(unixTimestamp)
  ) {
    unixTimestamp *= 1000;
  }

  return unixTimestamp as number;
};

/**
 * 将字符串类型的时间范围转换为Unix时间戳范围
 * @param timeRange 时间范围数组，包含起始时间和结束时间
 * @param format 格式化模板（可选）
 * @param isToSecondTimestamp 是否转换为秒级时间戳（可选，默认为false）
 * @param isToMillSecondTimestamp 是否转换为毫秒级时间戳（可选，默认为false）
 * @returns 对应的Unix时间戳范围
 */
export const convertTimeRangeToUnixTimestampRange = ({
  timeRange,
  format = DEFAULT_TIME_FORMAT_TEMPLATE,
  isToSecondTimestamp = false,
  isToMillSecondTimestamp = true,
}: {
  timeRange: string[] | number[] | undefined;
  format?: string;
  isToSecondTimestamp?: boolean;
  isToMillSecondTimestamp?: boolean;
}): number[] => {
  if (!timeRange || timeRange.length === 0) {
    return [];
  }
  const [startTime, endTime] = timeRange.map((value) =>
    canConvertToNumberFromCommon(value) ? Number(value) : value,
  );
  // 如果是['1640995200000', '1672531199999']，转成数值类型数组后直接返回
  if (isNumber(startTime) && isNumber(endTime)) {
    return [startTime, endTime];
  }
  const convertedStartTime = convertToUnixTimestamp({
    time: startTime,
    format,
    isToSecondTimestamp,
    isToMillSecondTimestamp,
  });
  const convertedEndTime = convertToUnixTimestamp({
    time: endTime,
    format,
    isToSecondTimestamp,
    isToMillSecondTimestamp,
  });
  return [convertedStartTime, convertedEndTime];
};

/**
 * 将对象中指定的字段转换为时间戳并返回转换后的对象
 * @param fields 包含字段和对应值的对象
 * @returns 转换后的对象
 */
export const convertFieldsToTimestamp = <T>({
  fields,
  target,
  convertProps = {},
}: {
  fields: Array<string>;
  target: any;
  convertProps?: Omit<ConvertToUnixTimestampProps, 'time'>;
}): T => {
  const convertedFields: Record<string, string | number> = pick(target, fields);

  fields.forEach((field) => {
    const value = convertedFields[field];
    convertedFields[field] = convertToUnixTimestamp({
      time: value,
      isToMillSecondTimestamp: true,
      ...convertProps,
    });
  });

  const result = { ...target, ...convertedFields } as T;
  return result;
};

File: veaiops/handler/errors/convert.py
# Copyright 2025 Beijing Volcano Engine Technology Co., Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This module handles the conversion of generic exceptions to HTTP exceptions."""

from typing import Callable, Dict, Type

from beanie.exceptions import DocumentNotFound
from bson.errors import InvalidId
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from pydantic import ValidationError
from pymongo.errors import DuplicateKeyError, NetworkTimeout, ServerSelectionTimeoutError

from veaiops.handler.errors import (
    BadRequestError,
    BaseHTTPExc,
    ConflictError,
    InternalServerError,
    RecordNotFoundError,
    ServerTimeout,
    ServiceUnavailableError,
)
from veaiops.utils.log import logger

# A mapping of exceptions to custom HTTP exceptions.
EXCEPTIONS_MAP: Dict[Type[Exception], Type[BaseHTTPExc]] = {
    DocumentNotFound: RecordNotFoundError,
    DuplicateKeyError: ConflictError,
    NetworkTimeout: ServerTimeout,
    ServerSelectionTimeoutError: ServiceUnavailableError,
    InvalidId: BadRequestError,
    ValidationError: BadRequestError,
    ValueError: BadRequestError,
}


def register_exceptions_handler(app: FastAPI) -> None:
    """Registers exception handlers for the FastAPI application.

    This function iterates over the EXCEPTIONS_MAP and registers a generic
    handler for each specified exception, mapping it to a corresponding
    HTTP exception.

    Args:
        app (FastAPI): The FastAPI application instance.
    """

    def make_handler(exc_cls: type[BaseHTTPExc]) -> Callable:
        """Creates an exception handler for a given HTTP exception class.

        Args:
            exc_cls (type[BaseHTTPExc]): The class of the HTTP exception to be handled.

        Returns:
            Callable: An async function that handles the exception and returns a JSONResponse.
        """

        async def handler(request: Request, exc: Exception) -> JSONResponse:
            """Handles the exception and converts it to a JSON response.

            Args:
                request (Request): The request object.
                exc (Exception): The exception instance.

            Returns:
                JSONResponse: A JSONResponse object with the appropriate status code and error message.
            """
            error_message = str(exc)
            logger.error(f"Catch exception in router {request.url.path}: {error_message}")
            return exc_cls(message=error_message).json_response()

        return handler

    for inner_exc_cls, api_exc_cls in EXCEPTIONS_MAP.items():
        app.exception_handler(inner_exc_cls)(make_handler(api_exc_cls))

    # register default exception handler
    app.exception_handler(Exception)(make_handler(InternalServerError))
